sdk/data/azcosmos/cosmos_location_cache.go

```diff
@@ -0,0 +1,307 @@
+	endpoint string
+	// props    map[string]interface{}
+}
+
+// AcctProperties represents a container for Azure Cosmos DB databases.
+type AcctProperties struct {
+	// The list of readable regions of this database account from the Azure Cosmos DB service.
+	readRegions []AcctRegion
+	// The list of writable regions of this database account from the Azure Cosmos DB service.
+	writeRegions []AcctRegion
+	// Indicates whether multiple write locations are enabled for this database account.
+	enableMultipleWriteLocations bool
+}
+
+type locationCache struct {
+	locationInfo              dbAcctLocationsInfo
+	defaultEndpt              url.URL
+	enableEndptDiscovery      bool
+	useMultipleWriteLocations bool
+	// connLimit                         int
+	rwMutex                           sync.RWMutex
+	locationUnavailabilityInfoMap     map[url.URL]locationUnavailabilityInfo
+	mapMutex                          sync.RWMutex
+	lastUpdateTime                    time.Time
+	enableMultipleWriteLocations      bool
+	unavailableLocationExpirationTime time.Duration
+}
+
+func newLocationCache(prefLocations []string, defaultEndpt url.URL) *locationCache {
+	return &locationCache{
+		defaultEndpt:                      defaultEndpt,
+		locationInfo:                      *newDbAcctLocationsInfo(prefLocations, defaultEndpt),
+		locationUnavailabilityInfoMap:     make(map[url.URL]locationUnavailabilityInfo),
+		unavailableLocationExpirationTime: defaultExpirationTime,
+	}
+}
+
+func (lc *locationCache) update(writeLocations []AcctRegion, readLocations []AcctRegion, prefList []string, enableMultipleWriteLocations bool) error {
+	lc.rwMutex.RLock()
+	nextLoc := copyDbAcctLocationsInfo(lc.locationInfo)
+	if prefList != nil {
+		nextLoc.prefLocations = prefList
+	}
+	lc.enableMultipleWriteLocations = enableMultipleWriteLocations
+	lc.refreshStaleEndpts()
+	if readLocations != nil {
+		availReadEndptsByLocation, availReadLocations, err := getEndptsByLocation(readLocations)
+		if err != nil {
+			lc.rwMutex.RUnlock()
+			return err
+		}
+		nextLoc.availReadEndptsByLocation = availReadEndptsByLocation
+		nextLoc.availReadLocations = availReadLocations
+	}
+
+	if writeLocations != nil {
+		availWriteEndptsByLocation, availWriteLocations, err := getEndptsByLocation(writeLocations)
+		if err != nil {
+			lc.rwMutex.RUnlock()
+			return err
+		}
+		nextLoc.availWriteEndptsByLocation = availWriteEndptsByLocation
+		nextLoc.availWriteLocations = availWriteLocations
+	}
+
+	nextLoc.writeEndpts = lc.getPrefAvailableEndpts(nextLoc.availWriteEndptsByLocation, nextLoc.availWriteLocations, write, lc.defaultEndpt)
+	nextLoc.readEndpts = lc.getPrefAvailableEndpts(nextLoc.availReadEndptsByLocation, nextLoc.availReadLocations, read, nextLoc.writeEndpts[0])
+	lc.lastUpdateTime = time.Now()
+	lc.rwMutex.RUnlock()
+	lc.rwMutex.Lock()
+	lc.locationInfo = nextLoc
+	lc.rwMutex.Unlock()
+	return nil
+}
+
+func (lc *locationCache) readEndpts() ([]url.URL, error) {
+	lc.mapMutex.RLock()
```

Comment:
Why do we need the lock on the read operation? https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos/src/Routing/LocationCache.cs#L92

Reply:
For the locks, these are the decisions I made:

- Utilize RWMutexes (Read-Write), which allow multiple read locks to be acquired or ONE Write lock concurrently.
- One RWmutex for the locationUnavailabilityInfoMap
- One RWmutex for all other values of the location cache. 
If we are not concerned with any race conditions for the things like the lastUpdateTime, we can remove some uses of the read locks.

Reply:
I would try to align with the current .NET implementation, if there are no locks there, I would assume none are needed?

Reply:
See the remaining open comment thread, for the current implementation, this is required. Based on the decision in the other thread, I will either revise or leave as is.